### PR TITLE
Tracker profiler package

### PR DESCRIPTION
**Pull request by Slava**
WIP.

- Expose references to computations in Tracker.
- Introduce a package called `tracker-profiler` that counts recomputations of every computation.
- Give Blaze autoruns descriptive names in the [`displayName`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/displayName) field.
- Spacebars generator should give more descriptive names to the views generated for helpers and other symbols (or identifier paths).

@avital

### CODE REVIEW

File: packages/blaze/builtins.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ Blaze.With = function (data, contentFunc) {
       // `data` is a reactive function
       view.autorun(function () {
         view.dataVar.set(data());
-      }, view.parentView);
+      }, view.parentView, ':setData');
```
Comment (stubailo) [inline packages/blaze/builtins.js]: Put the colon in when you join the strings

File: packages/blaze/view.js [hunk 2]
```diff
@@ -174,18 +174,27 @@ Blaze.View.prototype.autorun = function (f, _inViewScope) {
     throw new Error("Can't call View#autorun from a Tracker Computation; try calling it from the created or rendered callback");
   }
 
-  var templateInstanceFunc = Blaze.Template._currentTemplateInstanceFunc;
+  // Use a local dictionary for locals to allocate them on heap. Helps to avoid
+  // the stack overflow.
```
Comment (stubailo) [inline packages/blaze/view.js]: Explain this more - maybe there is a link you can add to the problem?

**Comment by stubailo** [comment]
Make tracker-profiler a separate repo or something - can we make a meteor/tools repo or something (not to be confused with meteor/meteor/tools :] )

**Comment by RobertLowe** [comment]
> Make tracker-profiler a separate repo or something - can we make a meteor/tools repo or something (not to be confused with meteor/meteor/tools :] )

This hints at the larger problem that core-packages must die. ;)

---

Is MDG supporting `tracker-profiler`? If not then why even make the modifications to `blaze` & `spacebars-compiler`. 

I'd like to see development tools supported (`tracker-profiler`, etc), which is a point for keeping it in core.

Just my two cents

:+1:

**Comment by stubailo** [comment]
@RobertLowe I'm not sure what your point is - should core packages die or not die?

I think we should have some "core packages" that are part of the bulletproof foundation of Meteor, but we should try to limit that number of packages to the minimum so that we can provide first-class support for all of them.

**Comment by RobertLowe** [comment]
> I'm not sure what your point is - should core packages die or not die?

OT: In my opinion, I find it odd that meteor has `core-packages` which aren't pulled from the dependancy management system (atmosphere). I understand there may be an array of reasons as why it is like that right now. `core-packages` act more as half-standard-lib half-mdg-supported packages. I think it would be better if the `core-packages` were split up into their own repos (long term), but that's just my two cents from the peanut gallery.

---

What I'm trying to point out: is that if you think `tracker-profiler` shouldn't be a part of the `core-packages` then I'm left to assume it's not going to have first-class support, but since `blaze` and `spacebars-compiler` have first-class support, why accept pulls for supporting something that isn't first-class in support. 

I mean, I think until the core can be more easily broken up, it should be part of core or the pull shouldn't be accepted, but that's just my two cents again.

File: packages/tracker/tracker.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
-//////////////////////////////////////////////////
+/////////////////////////////////////////////////////
 // Package docs at http://docs.meteor.com/#tracker //
-//////////////////////////////////////////////////
+/////////////////////////////////////////////////////
```
Comment (avital) [inline packages/tracker/tracker.js]: Nice

File: packages/tracker/tracker.js [hunk 4]
```diff
@@ -26,6 +26,9 @@ Tracker.active = false;
  */
 Tracker.currentComputation = null;
 
+// References to all computations created within the Tracker by id.
```
Comment (avital) [inline packages/tracker/tracker.js]: Add an explanation for what this is used for.

File: packages/blaze/view.js (hunk 2 above)
Comment (avital) [inline packages/blaze/view.js]: Yeah, write a short essay describing it to someone who's never thought of this before. ("Beginner's mind")

**Comment by avital** [comment]
Cool! Fix the comments and merge it.

**Comment by Slava** [comment]
@RobertLowe the changes proposed here are useful in general case, not just for the package that I built last sunday.

Giving names to Blaze functions make better stack-traces, better errors, etc. Changes to Tracker don't expose a particular API optimized for TrackerProfiler, but rather just exposes computations in an object. If you look at my code for the profiler, it monkey-patches the computations.

tracker-profiler is not going to have first class support, at least so far. But these changes open up possibilities for other packages built by Meteor community.

**Comment by stubailo** [comment]
@RobertLowe if you would like to build a certain package and there are some hooks/features missing from core Meteor that would enable it, you should submit a PR to add those!

**Pull request merged**